File: Grind75_1/week1/TwoSum.py
"""
https://leetcode.com/problems/two-sum/
"""

from typing import List

class Solution:

    # O(n)
    def twoSum(self, nums: List[int], target: int) -> List[int]:

        # dictionary to contain the iterated values from array 'nums'. Value as the 'key' and index as the 'value'.
        dictNums = {}
        # to assign the difference value that will be looked for in dictNums
        iDiff = None

        # iterate the array 'nums' and find the indexes of the values that equal to the target 'target'. Appends to dictNums as it goes.
        for x in range(0, len(nums)):
            # Calculate the value that would satisfy the target
            iDiff = target - nums[x]
            if (dictNums.get(iDiff) is not None):
                # if the diff value exists in the dictNums, return the current index and the 'value' of the diff value in dictNums
                return [x, dictNums.get(iDiff)]
            else:
                # add the arr value as the 'key' and the index as the 'value'.
                # Since 2 sum, if multiple instances of same number it is OK to overwrite the index in the 'value'.
                dictNums[nums[x]] = x

        return []
    
    # Brute force, O(n^2)
    def twoSumBrute(self, nums: List[int], target: int) -> List[int]:

        res = []

        iDiff = None
        iCurrIdx = 0

        iLenNums = len(nums)

        for x in range(0, iLenNums):
            iDiff = target - nums[x]
            if (iDiff == nums[x] and nums.count(iDiff) == 1):
                # must be the same index
                pass
            elif (nums.count(iDiff) > 0):
                # find the index of the 2nd number that will sum to the target
                for y in range(0, iLenNums):
                    if (x == y):
                        pass
                    elif (nums[y] == iDiff):
                        res.append(x)
                        res.append(y)
                        return res
            else:
                pass    
        return res

  
    

File: Grind75/Arrays/two-sum.js
// https://leetcode.com/problems/two-sum/description/

/*
* Naive method:
iterate i in nums
    iterate j in nums
        if nums[i] + nums[j] === target
            return [i, j]

- Time: O(n^2)
- Space: O(1)

* Map method
create a Map for:
    key: the diff needed for the index to equal the target
    val: the index of the number that needs the diff

iterate i in nums
    if (Map has nums[i]):
        the combination found
        return [Map.get(nums[i]), i]
    else
        diff = target - nums[i]
        Map.set(diff, i)

return [-1, -1]

- Time: O(n)
- Space: O(n)
*/

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function(nums, target) {
    if (nums.length < 2) {
        return [-1, -1]
    }

    const diffMap = new Map()
    for (let i = 0; i < nums.length; i ++) {
        if (diffMap.has(nums[i])) {
            return [diffMap.get(nums[i]), i]
        } else {
            diffMap.set(target - nums[i], i)
        }
    }

    return [-1, -1]
};